### PR TITLE
Ent - upsertPackageIDDoNothing vs upsertPackageIDIgnore

### PR DESCRIPTION
**Pull request by mrizzi**
# Description of the PR

PR meant to discuss and decide the approach to adopt for Ent upserts (https://github.com/guacsec/guac/issues/1198)

@ivanvanderbyl `upsertPackageIDDoNothing` and `upsertPackageIDIgnore` have been implemented as you proposed (or at least this was my intention :sweat_smile: )
I've run some tests and, as reported, they performed the same (https://github.com/guacsec/guac/issues/1198#issuecomment-1715949772)

TBH I feel like I misunderstood when expecting `upsertPackageIDDoNothing` to perform better: I thought during first time package ingestion.
BUT I noticed that ingesting the same SBOM multiple times, gave a better performance for `upsertPackageIDDoNothing`.

To sum up, when dealing with a package ingestion for the:

- 1st time, they perform the same for inserting
- n-th time,  `upsertPackageIDDoNothing` performs 3x faster than `upsertPackageIDIgnore` 

Is this in line with your expectations?
Could we declare  `upsertPackageIDDoNothing` the winner?

# PR Checklist

- [X] All commits have [a Developer Certificate of Origin (DCO)](https://wiki.linuxfoundation.org/dco) -- they are generated using `-s` flag to `git commit`.
- [ ] All new changes are covered by tests
- [ ] If GraphQL schema is changed, `make generate` has been run
- [ ] If `collectsub` protobuf has been changed, `make proto` has been run
- [ ] All CI checks are passing (tests and formatting)
- [ ] All dependent PRs have already been merged

### CODE REVIEW

**Comment by ivanvanderbyl** [comment]
Thanks for investigating @mrizzi — that is indeed in line with my expectations around performance. The reasoning is:

1. On first insert using `DO NOTHING`, the db has to still insert the record, so writing to disk, creating an index entry etc.
2. On second using `DO NOTHING` it hits the unique index (very fast) causing a conflict error to be thrown in the DB. However, DoNothing then throws the insert job away.
3. Querying for the record to get the ID is super quick since it only hits the index.

File: pkg/assembler/backends/ent/backend/package.go [hunk 1]
```diff
@@ -130,37 +130,72 @@ func (b *EntBackend) IngestPackage(ctx context.Context, pkg model.PkgInputSpec)
...
-		OnConflict(sql.ConflictColumns(packagenamespace.FieldNamespace, packagenamespace.FieldPackageID)).UpdateNewValues().ID(ctx)
+	nsID, err := client.PackageNamespace.Create().
+		SetPackageID(pkgID).
+		SetNamespace(valueOrDefault(pkg.Namespace, "")).
+		OnConflict(sql.ConflictColumns(packagenamespace.FieldNamespace, packagenamespace.FieldPackageID)).
+		DoNothing().
+		ID(ctx)
 	if err != nil {
-		return nil, errors.Wrap(err, "upsert package namespace")
+		if err != stdsql.ErrNoRows {
```
[the remark this answers is not in the document]
Comment (mrizzi) [reply pkg/assembler/backends/ent/backend/package.go:163]: This was meant to be a kind of optimization: if the error is `ErrNoRows` (i.e. there has been a conflict and `DoNothing` returned no ID, due to https://github.com/ent/ent/issues/1821) then it's the only case when it's worth running the query to get the `nsID`.

**Comment by ivanvanderbyl** [comment]
@mrizzi I took a closer look locally, I now properly understand your thinking around handling the NoRows error. LGTM, ship it.

**Review by mihaimaruseac — APPROVED** [review]

**Review by jeffmendoza — APPROVED** [review]
Thanks for the testing and improvement!

**Pull request merged**